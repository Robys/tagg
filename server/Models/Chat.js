const {Schema,model} = require('mongoose')

const ChatSchema = Schema({
    user: {type:Schema.Types.ObjectId,ref:"User"},
    contact: {type:Schema.Types.ObjectId,ref:"User"},
    messages: [{type:Schema.Types.ObjectId,ref:"Message"}],

})

const CHAT = model('Chat',ChatSchema)

module.exports = {CHAT}