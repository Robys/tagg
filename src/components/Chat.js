import {useState,useEffect} from 'react'
import {List} from '@material-ui/core'
import {Skeleton} from '@material-ui/lab'
import {CHAT} from '../api'
import Message from './Message'

/** The Chat in these application is the list of messages between
 * two users 
 * @constructor
 * @param {Object} sender - person who's send the message 
 * @param {Object} receiver - person who's received the message
 */


function Chat ({sender,receiver}){
    
    const [messages,SetMessages] = useState()

    useEffect(()=>{
        const getMessages = async () =>{
            const response = await CHAT(sender,receiver)
            SetMessages(response)
        }

        getMessages()

    },[sender,receiver])

    return(
     <List style={{maxHeight: '400px', overflow: 'auto'}}>
        {messages 
        ? messages.map(item => <Message message={item} key={item.id} id="message"/>)
            : <ul style={{listStyle:"none", marginTop:"10px"}}>
            <li style={{display:"inline-flex"}}>
            <Skeleton animation="wave" variant="circle" width={40} height={40} />
            </li>
            <li>
                <Skeleton animation="wave" variant="rect" height={100} />
            </li>
            </ul>
        }
        </List>
    )

}

/**
 * overflow-y: scroll;
 * 
 * {messages ? 
        <List style={{margin:"10px",height:"100%"}}>
            {messages.map(item => 
                <Message item={item} key={item.id}/>
                )}
        </List>
        : 
        <ul style={{listStyle:"none", marginTop:"10px"}}>
            <li style={{display:"inline-flex"}}>
            <Skeleton animation="wave" variant="circle" width={40} height={40} />
            </li>
            <li>
                <Skeleton animation="wave" variant="rect" height={100} />
            </li>
            </ul>
        }
 */

export default Chat