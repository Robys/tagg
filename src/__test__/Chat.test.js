import React from 'react'
import {act} from 'react-dom/test-utils'
import {shallow} from '../enzyme'
import Chat from '../components/Chat'

const sender = {
    id:"60edb9b0b291b127e48a8778",
}

const receiver = {
    id:"60f051ab648d18294c24c343",
}


const whenStable = async () =>
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve, 0));
  });

describe("testing <Chat/>",()=>{

    test("fetching messages", async ()=>{
        const container = shallow(<Chat sender={sender} receiver={receiver}/>)
        //expect(container.exists("#message")).toBe(false)
        await whenStable()
        expect(container.find("#message")).toBeTruthy()
        expect(container).toMatchSnapshot()

    })

})