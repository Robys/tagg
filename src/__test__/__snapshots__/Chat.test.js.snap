// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <Chat/> fetching messages 1`] = `
<WithStyles(ForwardRef(List))
  style={
    Object {
      "height": "100%",
      "margin": "10px",
    }
  }
>
  <ul
    style={
      Object {
        "listStyle": "none",
        "marginTop": "10px",
      }
    }
  >
    <li
      style={
        Object {
          "display": "inline-flex",
        }
      }
    >
      <WithStyles(ForwardRef(Skeleton))
        animation="wave"
        height={40}
        variant="circle"
        width={40}
      />
    </li>
    <li>
      <WithStyles(ForwardRef(Skeleton))
        animation="wave"
        height={100}
        variant="rect"
      />
    </li>
  </ul>
</WithStyles(ForwardRef(List))>
`;
